# The Neighborhoods of Toronto

## Part 1: Scraping Wikipedia

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = requests.get(url)

I use BeautifulSoup to parse the Wikipedia page and extract the table:

In [3]:
soup = BeautifulSoup(page.text, 'html.parser')
tbl = soup.find("table")

Then I import the table to a dataframe:

In [4]:
df = pd.read_html(str(tbl))[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Next, I drop all rows for which no Borough is assigned and reset the index. Also, I change the spelling in the column header from British English to American English:

In [5]:
index_names = df[ df['Borough'] == 'Not assigned' ].index 
  
df.drop(index_names, inplace = True) 
 
df.reset_index(drop=True, inplace=True)  

df = df.rename(columns={'Neighbourhood': 'Neighborhood'})

df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [6]:
df.shape

(103, 3)

The resulting table consists of 103 rows and 3 columns.

## Part 2: Geographical Coordinates of Toronto's Neighborhoods

In [7]:
filename = 'https://cocl.us/Geospatial_data'

Import the CSV file with the geographical coordinates into a dataframe:

In [8]:
df_coordinates = pd.read_csv(filename)
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Join the two tables on the column "Postal Code":

In [9]:
toronto = pd.merge(df, df_coordinates, on='Postal Code')
toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

One Postal Code in Downtown Toronto seems to be for PO Boxes. I remove it from the Dataframe. Another seems to be the a processing plant for business reply letters. I remove it as well: 

In [10]:
toronto.drop(df[df['Neighborhood'] == 'Stn A PO Boxes'].index, inplace = True)
toronto.drop(df[df['Neighborhood'] == 'Business reply mail Processing Centre, South Central Letter Processing Plant Toronto'].index, inplace = True)

In [11]:
toronto.shape

(101, 5)

There are 101 postal codes with neighborhoods in Toronto.

## Part 3: Analyzing the Neighborhoods

In [12]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import folium # map rendering library
import numpy as np # library to handle data in a vectorized manner
from sklearn.cluster import KMeans 

import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors

Let's focus on the inner city of Toronto, so I create a new dataframe with only the Boroughs that contain the word "Toronto":

In [13]:
inner_toronto = toronto[toronto.Borough.str.contains('toronto',case=False)]
inner_toronto.shape

(37, 5)

### Visualize the Neighborhoods

I use the geopy library to get the latitude and longitude values of Toronto:

In [14]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


I create a map of Toronto's inner city with the neignborhoods I found in the previous part:


In [15]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(inner_toronto['Latitude'], inner_toronto['Longitude'], inner_toronto['Borough'], inner_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Explore the inner city

Get started with Foursquare:

In [16]:
CLIENT_ID = 'ES24LSNJRRLUCV0GVWEF2R2JSB0K1HX2PL1PHNPIWQERDGOB' # your Foursquare ID
CLIENT_SECRET = 'FFNEJ241EVK3RPUHXL5CHCWZDSLDSD4ROUYCI1ZSIRK42ZC4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ES24LSNJRRLUCV0GVWEF2R2JSB0K1HX2PL1PHNPIWQERDGOB
CLIENT_SECRET:FFNEJ241EVK3RPUHXL5CHCWZDSLDSD4ROUYCI1ZSIRK42ZC4


Get the top 100 venues within 500 metres of every neighborhood and structure it into a dataframe:

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [18]:
toronto_venues = getNearbyVenues(names = inner_toronto['Neighborhood'], latitudes = inner_toronto['Latitude'], longitudes = inner_toronto['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [19]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [20]:
toronto_venues.shape

(1512, 7)

Foursquare found 1512 venues in the inner city of Toronto. 

In [21]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                                                         9   
Dufferin, Dovercourt Village                                            13   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest H

### Prepare for K-Means Clustering

In [22]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
neighborhood = toronto_onehot['Neighborhood']
toronto_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
toronto_onehot.insert(0, 'Neighborhood', neighborhood)

toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Theme Restaurant  Toy / Game Store  \
0                    0             0  ...                 0                 0   
1                    0             0  ...                 0                 0   
2                    0             0  ...                 0                 0   
3                    0             0  ...                 0                 0   
4                    0             0  ...                 0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  Yoga Studio  
0                      0         0              0            0  
1                      0         0              0            0  
2                      0         0              0            0  
3                      0         0              0            0  
4                      0         0              0            0  

[5 rows x 228 columns]

Next, I group rows by neighborhood and take the mean of the frequency of occurrence of each category


In [23]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
3                                 Central Bay Street                0.0   
4                                           Christie                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0   0.0000              0.0000        0.0000           0.000            0.000   
1   0.0000              0.0000        0.0000           0.000            0.000   
2   0.0625              0.0625        0.0625           0.125            0.125   
3   0.0000              0.0000        0.0000           0.000            0.000   
4   0.0000              0.0000        0.0000           0.000            0.000   

   Airport Terminal  American Restaurant  Antique Shop  ...  Theme Restaurant  \
0            0.0000                  0.0           0.0  ...               0.0   
1            0.0000                  0.0           0.0  ...               0.0   
2            0.0625                  0.0           0.0  ...               0.0   
3            0.0000                  0.0           0.0  ...               0.0   
4            0.0000                  0.0           0.0  ...               0.0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0    0.0            0.0                       0.018182   
1               0.0    0.0            0.0                       0.000000   
2               0.0    0.0            0.0                       0.000000   
3               0.0    0.0            0.0                       0.014706   
4               0.0    0.0            0.0                       0.000000   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0               0.0                    0.0  0.000000            0.0   
1               0.0                    0.0  0.000000            0.0   
2               0.0                    0.0  0.000000            0.0   
3               0.0                    0.0  0.014706            0.0   
4               0.0                    0.0  0.000000            0.0   

   Yoga Studio  
0     0.000000  
1     0.000000  
2     0.000000  
3     0.014706  
4     0.000000  

[5 rows x 228 columns]

In [24]:
toronto_grouped.shape

(37, 228)

Let's print each neighborhood along with the top 5 most common venues:


In [25]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1            Beer Bar  0.04
2        Cocktail Bar  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1       Nightclub  0.09
2     Coffee Shop  0.09
3  Breakfast Spot  0.09
4   Grocery Store  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0       Airport Lounge  0.12
1      Airport Service  0.12
2  Rental Car Location  0.06
3             Boutique  0.06
4          Coffee Shop  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1                Café  0.06
2      Sandwich Place  0.04
3  Italian Restaurant  0.04
4     Bubble Tea Shop  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3    Coffee Sh

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Here is a new dataframe that displays the top 10 venues by category for each neighborhood:


In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market              Beer Bar          Cocktail Bar   
1        Breakfast Spot             Nightclub           Coffee Shop   
2       Airport Service              Boutique       Harbor / Marina   
3                  Café        Sandwich Place    Italian Restaurant   
4                  Café                  Park    Athletics & Sports   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery           Cheese Shop    Seafood Restaurant   
1          Climbing Gym         Burrito Place            Restaurant   
2         Boat or Ferry   Rental Car Location                   Bar   
3      Department Store           Salad Place       Thai Restaurant   
4    Italian Restaurant            Restaurant            Baby Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Grocery Store                    Pub  
1    Italian Restaurant          Intersection                Stadium  
2           Coffee Shop                 Plane       Sculpture Garden  
3   Japanese Restaurant          Burger Joint        Bubble Tea Shop  
4           Candy Store             Nightclub            Coffee Shop

### Run K-Means Clustering

In [28]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters = kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 3, 3, 0, 3, 3, 3, 3, 0], dtype=int32)

Next, I create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood:


In [29]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = inner_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()


Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               3           Coffee Shop   
4   43.662301 -79.389494               3           Coffee Shop   
9   43.657162 -79.378937               3        Clothing Store   
15  43.651494 -79.375418               3           Coffee Shop   
19  43.676357 -79.293031               3                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                 Bakery                   Pub                  Park   
4            Yoga Studio            Hobby Shop        Sandwich Place   
9            Coffee Shop                  Café   Japanese Restaurant   
15                  Café          Cocktail Bar            Restaurant   
19     Health Food Store                   Pub           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
2         Breakfast Spot                  Café                      Theater   
4    Distribution Center            Restaurant                        Diner   
9        Bubble Tea Shop        Cosmetics Shop             Ramen Restaurant   
15             Gastropub              Beer Bar          American Restaurant   
19      Department Store     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
2           Dessert Shop   Distribution Center             Chocolate Shop  
4     Italian Restaurant                  Café              Smoothie Shop  
9              Bookstore           Pizza Place  Middle Eastern Restaurant  
15        Cosmetics Shop   Japanese Restaurant         Seafood Restaurant  
19   Dumpling Restaurant            Donut Shop           Doner Restaurant

Finally, I visualize the neighborhoods according to their cluster:

In [34]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine the Clusters

#### Cluster 1

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
25                        Christie               0         Grocery Store   
31    Dufferin, Dovercourt Village               0              Pharmacy   
69   High Park, The Junction South               0                   Bar   
80  University of Toronto, Harbord               0                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
25                  Café                  Park    Athletics & Sports   
31                Bakery         Grocery Store           Supermarket   
69                  Café       Thai Restaurant    Mexican Restaurant   
80             Bookstore                   Bar   Japanese Restaurant   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
25         Italian Restaurant            Restaurant            Baby Store   
31  Middle Eastern Restaurant           Music Venue                  Café   
69                     Bakery             Speakeasy             Bookstore   
80             Sandwich Place                Bakery           Yoga Studio   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
25                Candy Store             Nightclub            Coffee Shop  
31                    Brewery                   Bar                   Bank  
69  Cajun / Creole Restaurant           Music Venue          Grocery Store  
80                   Beer Bar            Beer Store             Restaurant

#### Cluster 2

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
91     Rosedale               1                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
91            Playground          Dance Studio  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
91   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
91               Dog Run    Distribution Center

#### Cluster 3

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
62     Roselawn               2           Music Venue                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
62           Yoga Studio      Department Store     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
62  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
62      Doner Restaurant                Dog Run

#### Cluster 4

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
2                           Regent Park, Harbourfront               3   
4         Queen's Park, Ontario Provincial Government               3   
9                            Garden District, Ryerson               3   
15                                     St. James Town               3   
19                                        The Beaches               3   
20                                        Berczy Park               3   
24                                 Central Bay Street               3   
30                           Richmond, Adelaide, King               3   
36  Harbourfront East, Union Station, Toronto Islands               3   
37                           Little Portugal, Trinity               3   
41                       The Danforth West, Riverdale               3   
42           Toronto Dominion Centre, Design Exchange               3   
43       Brockton, Parkdale Village, Exhibition Place               3   
47                     India Bazaar, The Beaches West               3   
48                     Commerce Court, Victoria Hotel               3   
54                                    Studio District               3   
67                                   Davisville North               3   
73                  North Toronto West, Lawrence Park               3   
74                The Annex, North Midtown, Yorkville               3   
75                             Parkdale, Roncesvalles               3   
79                                         Davisville               3   
81                                 Runnymede, Swansea               3   
84          Kensington Market, Chinatown, Grange Park               3   
86  Summerhill West, Rathnelly, South Hill, Forest...               3   
87  CN Tower, King and Spadina, Railway Lands, Har...               3   
96                        St. James Town, Cabbagetown               3   
97             First Canadian Place, Underground city               3   
99                               Church and Wellesley               3   

            1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
2                     Coffee Shop                Bakery                   Pub   
4                     Coffee Shop           Yoga Studio            Hobby Shop   
9                  Clothing Store           Coffee Shop                  Café   
15                    Coffee Shop                  Café          Cocktail Bar   
19                          Trail     Health Food Store                   Pub   
20                    Coffee Shop        Farmers Market              Beer Bar   
24                    Coffee Shop                  Café        Sandwich Place   
30                    Coffee Shop                  Café                 Hotel   
36                    Coffee Shop              Aquarium                 Hotel   
37                            Bar           Coffee Shop            Restaurant   
41               Greek Restaurant           Coffee Shop    Italian Restaurant   
42                    Coffee Shop                 Hotel            Restaurant   
43                           Café        Breakfast Spot             Nightclub   
47                           Park      Sushi Restaurant     Fish & Chips Shop   
48                    Coffee Shop            Restaurant                  Café   
54                    Coffee Shop             Gastropub                Bakery   
67                           Park                 Hotel        Breakfast Spot   
73                    Coffee Shop        Clothing Store           Yoga Studio   
74                           Café        Sandwich Place           Coffee Shop   
75                      Gift Shop        Breakfast Spot          Dessert Shop   
79                    Pizza Place          Dessert Shop        Sandwich Place   
81                    Coffee Shop                  Café      Sushi Restaurant   
84  Vegetarian / Vegan Restaurant    Mexican Restaurant     

#### Cluster 5

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
83  Moore Park, Summerhill East               4            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83                 Trail           Yoga Studio         Deli / Bodega   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
83      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 6

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
68  Forest Hill North & West, Forest Hill Road Park               5   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
68                  Park         Jewelry Store                 Trail   

   4th Most Common Venue 5th Most Common Venue        6th Most Common Venue  \
68      Sushi Restaurant         Deli / Bodega  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
68   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
68                Dog Run

#### Cluster 7

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
61  Lawrence Park               6                  Park              Bus Line   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
61           Swim School           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
61  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
61      Doner Restaurant                Dog Run